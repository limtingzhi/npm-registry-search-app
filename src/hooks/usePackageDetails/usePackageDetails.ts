import { useCallback, useEffect, useState } from 'react';
import { getPackageDetails } from '../../api/npmRegistry';
import { PackageObj } from '../../typings/npm-registry';

interface Props {
  packageName: string | null | undefined;
}

interface UsePackageDetails {
  errorMsg: string | null;
  isLoading: boolean;
  packageDetails: PackageObj | null;
}

function usePackageDetails(props: Props): UsePackageDetails {
  const { packageName } = props;

  const [isLoading, setIsLoading] = useState(false);
  const [errorMsg, setErrorMsg] = useState<string | null>(null);
  const [packageDetails, setPackageDetails] = useState<PackageObj | null>(null);

  const fetchPackageDetails = useCallback(async () => {
    setIsLoading(true);
    setErrorMsg(null);

    try {
      if (packageName == null) {
        setErrorMsg('Package name is missing.');
        return;
      }

      const result = await getPackageDetails(packageName);

      setPackageDetails(result);
    } catch (error: any) {
      setErrorMsg(error.message);
    } finally {
      setIsLoading(false);
    }
  }, [packageName]);

  useEffect(() => {
    fetchPackageDetails();
  }, [fetchPackageDetails]);

  return {
    errorMsg,
    isLoading,
    packageDetails,
  };
}

export default usePackageDetails;
