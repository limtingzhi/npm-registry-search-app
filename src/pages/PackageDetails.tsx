import { useParams } from 'react-router';

function PackageDetails() {
  const { name } = useParams<{ name: string; }>();

  return (
    <div>{`Package Details: ${name}`}</div>
  );
}

export default PackageDetails;
