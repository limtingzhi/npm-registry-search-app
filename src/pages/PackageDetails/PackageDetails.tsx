import { Alert, Container, Skeleton, Stack } from '@mui/material';
import { useParams } from 'react-router';
import Package from '../../components/Package/Package';
import usePackageDetails from '../../hooks/usePackageDetails/usePackageDetails';

function PackageDetails() {
  const { name } = useParams<{ name: string; }>();

  const { errorMsg, isLoading, packageDetails } = usePackageDetails({ packageName: name ?? null });

  return (
    <Container sx={{ paddingY: 3 }}>
      <Stack spacing={3}>
        {isLoading && (
          <Stack spacing={2}>
            <Skeleton height={100} />
            <Skeleton />
            <Skeleton />
          </Stack>
        )}
        {errorMsg && (
          <Alert severity="error" variant="filled">
            {errorMsg}
          </Alert>
        )}
        <Package packageDetails={packageDetails} />
      </Stack>
    </Container>
  );
}

export default PackageDetails;
