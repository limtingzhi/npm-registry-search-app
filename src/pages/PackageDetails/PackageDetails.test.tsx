/**
 * @jest-environment jsdom
 */

import { render } from '@testing-library/react';
import { MemoryRouter, Route, Routes } from 'react-router';
import { mockPackageObj } from '../../fixtures/packageDetails.fixture';
import usePackageDetails from '../../hooks/usePackageDetails/usePackageDetails';
import PackageDetails from './PackageDetails';

jest.mock('react-markdown', () => ({ children }: any) => (
  <div data-testid="react-markdown">{children}</div>
));
jest.mock('../../hooks/usePackageDetails/usePackageDetails');

describe('PackageDetails', () => {
  test('should match snapshot with default state', () => {
    const mockUsePackageDetails = {
      errorMsg: null,
      isLoading: false,
      packageDetails: mockPackageObj,
    };

    (usePackageDetails as jest.Mock).mockReturnValue(mockUsePackageDetails);

    const { asFragment } = render(
      <MemoryRouter initialEntries={['/package/react']}>
        <Routes>
          <Route path="/package/:name" element={<PackageDetails />} />
        </Routes>
      </MemoryRouter>,
    );

    expect(asFragment()).toMatchSnapshot();
  });

  test('should match snapshot with loading state', () => {
    const mockUsePackageDetails = {
      errorMsg: null,
      isLoading: true,
      packageDetails: null,
    };

    (usePackageDetails as jest.Mock).mockReturnValue(mockUsePackageDetails);

    const { asFragment } = render(
      <MemoryRouter initialEntries={['/package/react']}>
        <Routes>
          <Route path="/package/:name" element={<PackageDetails />} />
        </Routes>
      </MemoryRouter>,
    );

    expect(asFragment()).toMatchSnapshot();
  });

  test('should match snapshot with error message', () => {
    const mockUsePackageDetails = {
      errorMsg: 'Package not found',
      isLoading: false,
      packageDetails: null,
    };

    (usePackageDetails as jest.Mock).mockReturnValue(mockUsePackageDetails);

    const { asFragment } = render(
      <MemoryRouter initialEntries={['/package/react']}>
        <Routes>
          <Route path="/package/:name" element={<PackageDetails />} />
        </Routes>
      </MemoryRouter>,
    );

    expect(asFragment()).toMatchSnapshot();
  });
});
