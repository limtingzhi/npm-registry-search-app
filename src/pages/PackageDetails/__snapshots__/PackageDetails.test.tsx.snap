// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`PackageDetails should match snapshot with default state 1`] = `
<DocumentFragment>
  <div
    class="MuiContainer-root MuiContainer-maxWidthLg css-j81c45-MuiContainer-root"
  >
    <div
      class="MuiStack-root css-v2woy3-MuiStack-root"
    >
      <div
        class="MuiStack-root css-1sazv7p-MuiStack-root"
      >
        <h4
          class="MuiTypography-root MuiTypography-h4 MuiTypography-gutterBottom css-1it4a80-MuiTypography-root"
        >
          react
        </h4>
        <hr
          class="MuiDivider-root MuiDivider-fullWidth css-1jpc804-MuiDivider-root"
        />
        <div
          class="MuiGrid2-root MuiGrid2-container MuiGrid2-direction-xs-column MuiGrid2-direction-md-row-reverse MuiGrid2-spacing-xs-5 css-11psefe-MuiGrid2-root"
        >
          <div
            class="MuiGrid2-root MuiGrid2-container MuiGrid2-direction-xs-column MuiGrid2-grid-xs-12 MuiGrid2-grid-md-4 css-1lorcj6-MuiGrid2-root"
          >
            <div
              class="MuiBox-root css-0"
            >
              <h6
                class="MuiTypography-root MuiTypography-h6 css-1miy0lu-MuiTypography-root"
              >
                Author
              </h6>
              <p
                class="MuiTypography-root MuiTypography-body1 css-rizt0-MuiTypography-root"
              >
                Test Author
              </p>
            </div>
            <div
              class="MuiBox-root css-0"
            >
              <h6
                class="MuiTypography-root MuiTypography-h6 css-1miy0lu-MuiTypography-root"
              >
                Description
              </h6>
              <p
                class="MuiTypography-root MuiTypography-body1 css-rizt0-MuiTypography-root"
              >
                React is a JavaScript library for building user interfaces.
              </p>
            </div>
            <div
              class="MuiBox-root css-0"
            >
              <h6
                class="MuiTypography-root MuiTypography-h6 css-1miy0lu-MuiTypography-root"
              >
                Last Updated
              </h6>
              <p
                class="MuiTypography-root MuiTypography-body1 css-rizt0-MuiTypography-root"
              >
                15 Mar 2025, 12:20 am
              </p>
            </div>
            <div
              class="MuiBox-root css-0"
            >
              <h6
                class="MuiTypography-root MuiTypography-h6 css-1miy0lu-MuiTypography-root"
              >
                License
              </h6>
              <p
                class="MuiTypography-root MuiTypography-body1 css-rizt0-MuiTypography-root"
              >
                MIT
              </p>
            </div>
            <div
              class="MuiBox-root css-0"
            >
              <h6
                class="MuiTypography-root MuiTypography-h6 css-1miy0lu-MuiTypography-root"
              >
                Version
              </h6>
              <p
                class="MuiTypography-root MuiTypography-body1 css-rizt0-MuiTypography-root"
              >
                19.0.0
              </p>
            </div>
          </div>
          <div
            class="MuiGrid2-root MuiGrid2-direction-xs-row MuiGrid2-grid-xs-12 MuiGrid2-grid-md-8 css-8boly6-MuiGrid2-root"
          >
            <div
              data-testid="react-markdown"
            >
              # Test README
            </div>
          </div>
        </div>
      </div>
    </div>
  </div>
</DocumentFragment>
`;

exports[`PackageDetails should match snapshot with error message 1`] = `
<DocumentFragment>
  <div
    class="MuiContainer-root MuiContainer-maxWidthLg css-j81c45-MuiContainer-root"
  >
    <div
      class="MuiStack-root css-v2woy3-MuiStack-root"
    >
      <div
        class="MuiPaper-root MuiPaper-elevation MuiPaper-rounded MuiPaper-elevation0 MuiAlert-root MuiAlert-colorError MuiAlert-filledError MuiAlert-filled css-1lw0bg-MuiPaper-root-MuiAlert-root"
        role="alert"
        style="--Paper-shadow: none;"
      >
        <div
          class="MuiAlert-icon css-vab54s-MuiAlert-icon"
        >
          <svg
            aria-hidden="true"
            class="MuiSvgIcon-root MuiSvgIcon-fontSizeInherit css-1ckov0h-MuiSvgIcon-root"
            data-testid="ErrorOutlineIcon"
            focusable="false"
            viewBox="0 0 24 24"
          >
            <path
              d="M11 15h2v2h-2zm0-8h2v6h-2zm.99-5C6.47 2 2 6.48 2 12s4.47 10 9.99 10C17.52 22 22 17.52 22 12S17.52 2 11.99 2zM12 20c-4.42 0-8-3.58-8-8s3.58-8 8-8 8 3.58 8 8-3.58 8-8 8z"
            />
          </svg>
        </div>
        <div
          class="MuiAlert-message css-zioonp-MuiAlert-message"
        >
          Package not found
        </div>
      </div>
    </div>
  </div>
</DocumentFragment>
`;

exports[`PackageDetails should match snapshot with loading state 1`] = `
<DocumentFragment>
  <div
    class="MuiContainer-root MuiContainer-maxWidthLg css-j81c45-MuiContainer-root"
  >
    <div
      class="MuiStack-root css-v2woy3-MuiStack-root"
    >
      <div
        class="MuiStack-root css-1sazv7p-MuiStack-root"
      >
        <span
          class="MuiSkeleton-root MuiSkeleton-text MuiSkeleton-pulse css-kkcbhd-MuiSkeleton-root"
          style="height: 100px;"
        />
        <span
          class="MuiSkeleton-root MuiSkeleton-text MuiSkeleton-pulse css-kkcbhd-MuiSkeleton-root"
        />
        <span
          class="MuiSkeleton-root MuiSkeleton-text MuiSkeleton-pulse css-kkcbhd-MuiSkeleton-root"
        />
      </div>
    </div>
  </div>
</DocumentFragment>
`;
